### PR TITLE
Add error handling for dynamic list in SDK

### PR DESCRIPTION
**Pull request by 16oeahr**
# Description

Add error handling for dynamic list in SDK.

# All Promptflow Contribution checklist:
- [X] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 1]
```diff
@@ -209,6 +209,23 @@ def load_python_module(content, source=None):
         ) from e
 
 
+def load_function_from_function_path(func_path: str):
```
Comment (chjinche) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:212]: Looks inappropriate to put in `tool_meta_generator`

File: src/promptflow/tests/executor/unittests/_core/test_tools_manager.py [hunk 2]
```diff
@@ -204,10 +245,197 @@ def test_collect_package_tools_and_connections(self, install_custom_tool_pkg):
...
+
+    @pytest.mark.parametrize(
+        "res",
+        [
+            (
+                [
+                    {
+                        "value": "fig0",
+                        "display_value": "My_fig0",
+                        "hyperlink": "https://www.google.com/search?q=fig0",
```
Comment (chjinche) [inline src/promptflow/tests/executor/unittests/_core/test_tools_manager.py:388]: plan to change all `google` to `bing`

File: src/promptflow/tests/executor/unittests/_core/test_tools_manager.py [hunk 3]
```diff
@@ -100,6 +106,41 @@ def sample_tool(input: str):
     return input
 
 
+def mock_func1():
```
Comment (chjinche) [inline src/promptflow/tests/executor/unittests/_core/test_tools_manager.py:109]: add some comments here to let other devs know this is for dynamic list function testing.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 1 above)
Comment (chjinche) [reply src/promptflow/promptflow/_core/tool_meta_generator.py:212]: Let's move all codes except `gen_dynamic_list` to `https://github.com/microsoft/promptflow/blob/dfbad48467a25dc0df424a0210f08a1c74a61732/src/promptflow/promptflow/_utils/tool_utils.py`

File: src/promptflow/promptflow/_core/tools_manager.py [hunk 4]
```diff
@@ -194,23 +200,54 @@ def append_workspace_triple_to_func_input_params(func_sig_params, func_input_par
...
+
+    The response is a list of items. Each item is a dict with the following keys:
+        - value: for backend use. Required.
+        - display_value: for UI display. Optional.
+        - hyperlink: external link. Optional.
+        - description: information icon tip. Optional.
+    The response can not be empty.
+    """
+    if not response:
+        raise ListFunctionResponseError(f"{f} response can not be empty.")
```
Comment (chjinche) [inline src/promptflow/promptflow/_core/tools_manager.py:214]: Let's improve the message to be more friendly to tool user who may be unfamiliar with the tool impl.

Comment (chjinche) [reply src/promptflow/promptflow/_core/tools_manager.py:214]: For example, "cannot list items due to xxx. Please contact tool author for trouble shooting."

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 5]
```diff
@@ -209,6 +209,23 @@ def load_python_module(content, source=None):
...
+    The function path should be in the format of "module_name.function_name".
+    """
+    try:
+        module_name, func_name = func_path.rsplit(".", 1)
+        module = importlib.import_module(module_name)
+        f = getattr(module, func_name)
+        if callable(f):
+            return f
+        else:
+            raise FunctionPathValidationError(f"'{f}' is not callable.")
```
Comment (chjinche) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:224]: Unable to display list of items due to XXX. Please contact the tool author/support team for troubleshooting assistance.

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 6]
```diff
@@ -209,6 +209,23 @@ def load_python_module(content, source=None):
...
+    except Exception as e:
+        raise FunctionPathValidationError(f"Failed to parse function from function path: '{func_path}' with error: {e}")
```
Comment (chjinche) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:226]: Unable to display list of items due to XXX. The list function needs to follow the format "module_name.function_name" to work properly. Please contact the tool author/support team for troubleshooting assistance.

Comment (16oeahr) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: updated

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 5 above)
Comment (16oeahr) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: updated

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 1 above)
Comment (16oeahr) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: moved

File: src/promptflow/tests/executor/unittests/_core/test_tools_manager.py (hunk 3 above)
Comment (16oeahr) [reply src/promptflow/tests/executor/unittests/_core/test_tools_manager.py]: added

File: src/promptflow/promptflow/_core/tools_manager.py (hunk 4 above)
Comment (16oeahr) [reply src/promptflow/promptflow/_core/tools_manager.py]: added

**Review by chjinche — APPROVED** [review]

**Pull request merged**